File: src/screens/PresentsScreen.lua
-- Skelly splash screen, "Taffer presents".
--
-- By Chris Herborth (https://github.com/Taffer)
-- MIT license, see LICENSE.md for details.

local Class = require 'lib/middleclass/middleclass'
local ScreenBase = require 'src/screens/ScreenBase'

local ColorFade = require 'src/ColorFade'
local ImageButton = require 'src/ui/ImageButton'
local Label = require 'src/ui/Label'
local UIBase = require 'src/ui/UIBase'

local FADE_TIME = 0.5 -- Set to 2.0 for production.

local PresentsScreen = Class('PresentsScreen', ScreenBase)
function PresentsScreen:initialize(resources, state)
    ScreenBase.initialize(self, resources, state)
    self:setNextScreen('Title')

    self.resources.text:setLanguage(gameState.settings:get('language'))

    self.resources.fonts.default_serif = love.graphics.newFont('graphics/A_Font_with_Serifs.ttf', 72)
    self.resources.fonts.default_mono = love.graphics.newFont('graphics/LiberationMono-Bold.ttf', 16)
    self.resources.images.love_logo = love.graphics.newImage('graphics/love-game-0.10.png')
    self.resources.images.taffer = love.graphics.newImage('graphics/taffer-ronos.png')
    self.resources.music.theme = love.audio.newSource('music/Heroic Demise (New).mp3',  'stream')
    love.audio.setVolume(gameState.settings:get('music_volume') * (gameState.settings:get('overall_volume')))
    love.audio.play(self.resources.music.theme) -- start playing ASAP

    local presents_text = self.resources.text:getText('presents')
    self.taffer_text = presents_text.taffer_text
    self.love_text = presents_text.love_text

    self.fade = ColorFade:new({0, 0, 0, 1}, {0, 0, 0, 0}, FADE_TIME)
    self.fade_out = false -- fade in first...
    self.exit_countdown = FADE_TIME * 2 -- seconds after fade to automatically exit

    local love_logo = self.resources.images.love_logo
    local logo_quad = love.graphics.newQuad(0, 0, love_logo:getWidth(), love_logo:getHeight(), love_logo)

    local taffer_logo = self.resources.images.taffer
    local taffer_quad = love.graphics.newQuad(0, 0, taffer_logo:getWidth(), taffer_logo:getHeight(), taffer_logo)

    self.ui = {
        Label:new(state.scr_width / 2, 16, self.taffer_text,
            self.resources.fonts.default_serif, {1, 1, 1, 1}, 'centre'),
        ImageButton:new((state.scr_width - taffer_logo:getWidth()) / 2, 120, taffer_logo, taffer_quad),
        ImageButton:new((state.scr_width - love_logo:getWidth()) / 2, 500, love_logo, logo_quad),
        Label:new(state.scr_width / 2, 640, self.love_text,
            self.resources.fonts.default_mono, {1, 1, 1, 1}, 'centre'),
    }
end

-- Render this screen's contents.
function PresentsScreen:draw()
    love.graphics.clear(0, 0, 0, 1)

    for i in ipairs(self.ui) do
        self.ui[i]:draw()
    end

    love.graphics.setColor(unpack(self.fade:getColor()))
    love.graphics.rectangle('fill', 0, 0, gameState.scr_width, gameState.scr_height)
end

-- Update the screen.
function PresentsScreen:update(dt)
    self.fade:update(dt)

    if self.fade_out then
        -- If we're fading out...
        if self.fade:isDone() then
            self:setExit()
        end
    else
        -- If we're fading in...
        if self.fade:isDone() then
            self.exit_countdown = self.exit_countdown - dt
            if self.exit_countdown < 0 then
                self.fade = ColorFade:new({0, 0, 0, 0}, {0, 0, 0, 1}, FADE_TIME)
                self.fade_out = true
            end
        end
    end
end

-- Check for input events.
function PresentsScreen:checkInputs(keyboard, mouse, gamepad)
    if keyboard['escape'] or mouse[1] or gamepad['a'] then
        self:setExit()
    end
end

return PresentsScreen

File: src/screens/PresentsScreen.py
''' Skelly "Taffer presents..." screen class.

By Chris Herborth (https://github.com/Taffer)
MIT license, see LICENSE.md for details.
'''

import pygame

from .ScreenBase import ScreenBase
from ..ui import ColorFade
from ..ui import ImageButton
from ..ui import Label

BLACK = pygame.Color('black')
BLACK_ALPHA = pygame.Color(BLACK.r, BLACK.g, BLACK.g, 0)  # BLACK, but fully transparent
WHITE = pygame.Color('white')


class PresentsScreen(ScreenBase):
    def __init__(self, game) -> None:
        super().__init__(game)

        self.next_screen = 'Title'
        '''
        self.resources.text:setLanguage(gameState.settings:get('language'))
        '''

        self.game.resources['fonts']['default_serif'] = pygame.freetype.Font('fonts/A_Font_with_Serifs.ttf', 72)
        self.game.resources['fonts']['default_mono'] = pygame.freetype.Font('fonts/LiberationMono-Bold.ttf', 16)
        self.game.resources['fonts']['germania'] = pygame.freetype.Font('fonts/GermaniaOne-Regular.ttf', 18)

        self.game.resources['images']['pygame_logo'] = pygame.image.load('graphics/pygame-logo.png').convert_alpha()
        self.game.resources['images']['taffer'] = pygame.image.load('graphics/taffer-ronos.png').convert_alpha()

        self.game.resources['music']['theme'] = pygame.mixer.music.load('music/Heroic Demise (New).ogg')

        pygame.mixer.music.set_volume(self.game.settings.get('music_volume') * self.game.settings.get('overall_volume'))
        pygame.mixer.music.play()

        presents_text = self.game.text.get_text('presents')
        self.taffer_text = presents_text['taffer_text']
        self.pygame_text = presents_text['pygame_text']

        self.fade = ColorFade(BLACK, BLACK_ALPHA, 1)  # 1 second fade
        self.fade_out = False
        self.exit_countdown = 2  # Seconds after fade to auto-exit.

        self.ui = []
        rect = self.game.resources['images']['taffer'].get_rect()
        self.ui.append(ImageButton((self.game.screen_width - rect.width) / 2, 120, self.game.resources['images']['taffer']))

        rect = self.game.resources['images']['pygame_logo'].get_rect()
        self.ui.append(ImageButton((self.game.screen_width - rect.width) / 2, 580, self.game.resources['images']['pygame_logo']))

        self.ui.append(Label(self.game.screen_width / 2, 16, self.taffer_text, self.game.resources['fonts']['default_serif'],
                       WHITE, 'centre'))

        self.ui.append(Label(self.game.screen_width / 2, 640, self.pygame_text, self.game.resources['fonts']['germania'],
                       WHITE, 'centre'))

    def draw(self) -> None:
        self.game.surface.fill(BLACK)

        for item in self.ui:
            item.draw()

        self.fade.draw()

    def update(self, dt: float) -> None:
        self.fade.update(dt)

        if self.fade_out:
            # If we're fading out...
            if self.fade.is_done():
                self.can_exit = True
        else:
            # If we're fading in...
            if self.fade.is_done():
                self.exit_countdown = self.exit_countdown - dt
                if self.exit_countdown < 0:
                    self.fade = ColorFade(BLACK_ALPHA, BLACK, 1)
                    self.fade_out = True
